text/3519-arbitrary-self-types-v2.md

```diff
@@ -0,0 +1,653 @@
+
+If you're implementing a smart pointer `P<T>`, and you need to allow `impl T { fn method(self: P<T>) { ... }}`, yet you can't allow a reference `&T` to exist, any option for implementing `Deref::deref` has drawbacks:
+
+* Specify `Deref::Target=T` and panic in `Deref::deref`. Not good.
+* Specify `Deref::Target=*const T`. This is only possible if your smart pointer type contains a `*const T` which you can reference - this isn't the case for (for instance) weak pointers or types containing `NonNull`.
+
+Therefore, the current Arbitrary Self Types v2 provides a separate `Receiver` trait, so that there's no need to provide an awkward `Deref::deref` implementation.
+
+This v2 version has two other differences relative to the existing unstable `arbitrary_self_type` feature:
+* We won't allow raw pointer receivers, yet. It's highly desirable that we do so in future - this is discussed under the [enable for pointers](#enable-for-pointers) section.
+* We will block generic receivers. See the [diagnostics section for reasoning](#diagnostics).
+
+Aside from these differences, Arbitrary Self Types v2 is similar to the existing unstable `arbitrary_self_types` feature.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When declaring a method, users can also declare the type of the `self` receiver to be any type `T` where `T: Receiver<Target = Self>`, in addition to using `Self` by value or reference.
+
+The `Receiver` trait is simple and only requires specifying the `Target` type:
+
+```rust
+trait Receiver {
+    type Target: ?Sized;
+}
+```
+
+The `Receiver` trait is already implemented for many standard library types:
+- smart pointers in the standard library: `Rc<Self>`, `Arc<Self>`, `Box<Self>`, and `Pin<SomeSmartPtr<Self>>` (and in fact, any type which implements `Deref`)
+- references: `&Self` and `&mut Self`
+- pointers: `*const Self` and `*mut Self`
+
+Shorthand exists for references, so that `self` with no ascription is of type `Self`, `&self` is of type `&Self` and `&mut self` is of type `&mut Self`.
+
+All of the following self types are valid:
+
+```rust
+impl Foo {
+    fn by_value(self /* self: Self */);
+    fn by_ref(&self /* self: &Self */);
+    fn by_ref_mut(&mut self /* self: &mut Self */);
+    fn by_ptr(self: *const Self);
+    fn by_mut_ptr(self: *mut Self);
+    fn by_box(self: Box<Self>);
+    fn by_rc(self: Rc<Self>);
+    fn by_custom_ptr(self: CustomPtr<Self>);
+}
+
+struct CustomPtr<T>(*const T);
+
+impl<T> Receiver for CustomPtr<T> {
+    type Target = T;
+}
+```
+
+## Recursive arbitrary receivers
+
+Receivers are recursive and therefore allowed to be nested. If type `T` implements `Receiver<Target=U>`, and type `U` implements `Receiver<Target=Self>`, `T` is a valid receiver (and so on outward). This is the behavior for the current special-cased self types (`Pin`, `Box` etc.) so as we remove the special-casing we need to retain this property.
```

Comment:
I found the last sentence here hard to read, I initially read it as:
- so as we remove, the special casing we need to retain this property _(missing words?)_.
- so ~~as~~ we remove the special casing we need to retain this property.

The third time through, I realised the phrasing was meant to be:
```suggestion
Receivers are recursive and therefore allowed to be nested. If type `T` implements `Receiver<Target=U>`, and type `U` implements `Receiver<Target=Self>`, `T` is a valid receiver (and so on outward). This is the behavior for the current special-cased self types (`Pin`, `Box` etc.), so as we remove the special-casing, we need to retain this property.
```

Another alternative is to split the long sentence:
```suggestion
Receivers are recursive and therefore allowed to be nested. If type `T` implements `Receiver<Target=U>`, and type `U` implements `Receiver<Target=Self>`, `T` is a valid receiver (and so on outward). This is the behavior for the current special-cased self types (`Pin`, `Box` etc.). So as we remove the special-casing, we need to retain this property.
```